src/storage/src/source/postgres/replication.rs

```diff
@@ -33,6 +33,41 @@
 //!       │ updates      │ output
 //!       v              v
 //! ```
+//!
+//! # Progress tracking
+//!
+//! In order avoid causing excessive resource usage in the upstream server it's important to
+//! track the LSN that we have successfully committed to persist and communicate that back to
+//! PostgreSQL. Under normal operation this gauge of progress is provided by the presence of
+//! transactions themselves. Since at a given LSN offset there can be only a single message, when a
+//! transaction is received and processed we can infer that we have seen all the messages that are
+//! not beyond `commit_lsn + 1`.
+//!
+//! Things are a bit more complicated in the absence of transcations though because even though we
+//! don't receive any the server might very well be generating WAL records. This can happen if
+//! there is a separate logical database performing writes (which is the case for RDS databases),
+//! or, in servers running PostgreSQL version 15 or greater, the logical replication process
+//! includes an optimization that omits empty transactions, which can happen if you're only
+//! replicating a subset of the tables and there writes going to the other ones.
+//!
+//! If we fail to detect this situation and don't send LSN feedback in a timely manner the server
+//! will be forced to keep around WAL data that can eventually lead to disk space exhaustion.
+//!
+//! In the absence of transactions the only available piece of information in the replication
+//! stream are keepalive messages. Keepalive messages are documented[1] to contain the current end
+//! of WAL on the server. That is a useless number when it comes to progress tracking because there
+//! might be pending messages at LSNs between the last received commit_lsn and the current end of
+//! WAL.
+//!
+//! Fortunatelly for us, the documentation for PrimaryKeepalive messages is wrong and it actually
+//! contains the last *sent* LSN[2]. Here sent doesn't necessarily mean sent over the wire, but sent
+//! to the upstream process that is handling producing the logical stream. Therefore, if we receive
+//! a keepalive with a particular LSN we can be certain that there are no other replication
+//! messages at previous LSNs, because they would have been already generated and received. We
+//! therefore connect the keepalive messages directly to our capability.
```

Comment:
can we add a TODO to fix the documentation upstream? I would hate for it to justify them changing this behavior in the future...

Reply:
I am working on this unless Petros beats me to it